pkg/controller/networkpolicy/clusternetworkpolicy.go

```diff
@@ -255,6 +260,91 @@ func (n *NetworkPolicyController) deleteNamespace(old interface{}) {
 	}
 }
 
+func (c *NetworkPolicyController) filterAGsFromNodeLabels(node *v1.Node) sets.String {
+	ags := sets.NewString()
+	addressGroupObjs, _ := c.addressGroupStore.GetByIndex(store.IsNodeAddressGroupIndex, store.IsNodeAddressGroupIndex)
+	for _, addressGroupObj := range addressGroupObjs {
+		addressGroup := addressGroupObj.(*antreatypes.AddressGroup)
+		nodeSelector := addressGroup.Selector.NodeSelector
+		if nodeSelector.Matches(labels.Set(node.GetLabels())) {
+			ags.Insert(addressGroup.Name)
+		}
+	}
+	return ags
+}
+
+func (c *NetworkPolicyController) addNode(obj interface{}) {
+	node := obj.(*v1.Node)
+	affectedAGs := c.filterAGsFromNodeLabels(node)
+	for _, key := range affectedAGs.List() {
+		c.enqueueAddressGroup(key)
+	}
+	klog.V(2).InfoS("Processed Node CREATE event", "nodeName", node.Name, "affectedAGs", affectedAGs.Len())
+}
+
+func (c *NetworkPolicyController) deleteNode(obj interface{}) {
+	node, ok := obj.(*v1.Node)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.ErrorS(nil, "Processed Node DELETE event error", "obj", obj)
+			return
+		}
+		node, ok = tombstone.Obj.(*v1.Node)
+		if !ok {
+			klog.ErrorS(nil, "Processed Node DELETE event error", "obj", tombstone.Obj)
+			return
+		}
+	}
+	// enqueue affected address group
+	affectedAGs := c.filterAGsFromNodeLabels(node)
+	for _, key := range affectedAGs.List() {
+		c.enqueueAddressGroup(key)
+	}
+	klog.V(2).InfoS("Processed Node DELETE event", "nodeName", node.Name, "affectedAGs", affectedAGs.Len())
+}
+
+func nodeIPChanged(oldNode, newNode *v1.Node) (changed bool, err error) {
+	var oldIPs, newIPs, oldNodeGWIPs, newNodeGWIPs *ip.DualStackIPs
+	if oldIPs, err = k8s.GetNodeAddrs(oldNode); err != nil {
+		return
+	}
+	if newIPs, err = k8s.GetNodeAddrs(newNode); err != nil {
+		return
+	}
+	if oldNodeGWIPs, err = k8s.GetNodeAddressFromAnnotations(oldNode, types.NodeAntreaGWAddressAnnotationKey); err != nil {
+		return
+	}
+	if newNodeGWIPs, err = k8s.GetNodeAddressFromAnnotations(newNode, types.NodeAntreaGWAddressAnnotationKey); err != nil {
+		return
+	}
+	return !reflect.DeepEqual(newIPs, oldIPs) || !reflect.DeepEqual(oldNodeGWIPs, newNodeGWIPs), nil
+}
+
+func (c *NetworkPolicyController) updateNode(oldObj, newObj interface{}) {
+	node := newObj.(*v1.Node)
+	oldNode := oldObj.(*v1.Node)
+	ipChanged, err := nodeIPChanged(oldNode, node)
+	if err != nil {
+		klog.ErrorS(err, "Processed Node UPDATE error", "nodeName", node.Name)
+		return
+	}
+	if reflect.DeepEqual(node.GetLabels(), oldNode.GetLabels()) && !ipChanged {
+		klog.V(2).InfoS("Processed Node UPDATE event, labels not changed", "nodeName", node.Name)
+		return
+	}
+	oldMatches, newMatches := c.filterAGsFromNodeLabels(oldNode), c.filterAGsFromNodeLabels(node)
```

Comment:
This it not very efficient. What if the labels don't change?

It should compute labelsChanged and ipChanged first, then
```
if !labelsChanged && !ipChanged {
    return
}

affectedAGs := c.filterAGsFromNodeLabels(node)
if labelsChanged {
    oldAGs := c.filterAGsFromNodeLabels(oldNode)
    if ipChanged {
        affectedAGs = utilsets.MergeString(affectedAGs, oldAGs)
    } else {
        affectedAGs = utilsets.SymmetricDifferenceString(affectedAGs, oldAGs)
    }
}
for ag := affectedAGs {
...
}
```

Reply:
Updated.